# NIFTY50 Market Direction Prediction

## Notebook 01 — Data Collection

### Objective

This notebook is responsible for collecting, validating, and storing all raw datasets used in the project.

No preprocessing, feature engineering, or modeling is performed here.

## Workflow

1. Download historical market data
2. Verify data quality
3. Inspect columns and date ranges
4. Save raw datasets

In [ ]:
# Import libraries

import pandas as pd
import numpy as np

In [ ]:
import pandas as pd
import numpy as np
import yfinance as yf

In [ ]:
# Project constants

NIFTY_TICKER = "^NSEI"
DATA_FOLDER = "data/raw"

In [ ]:
print(pd.__version__)

2.2.2


In [ ]:
#!pip install yfinance


In [ ]:
import yfinance as yf

In [ ]:
# Download complete available history

nifty = yf.download(
    "^NSEI",
    period="max",
    auto_adjust=False
)

[*********************100%***********************]  1 of 1 completed


In [ ]:
nifty.head()
nifty.tail()
nifty.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4621 entries, 2007-09-17 to 2026-07-21
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (Adj Close, ^NSEI)  4621 non-null   float64
 1   (Close, ^NSEI)      4621 non-null   float64
 2   (High, ^NSEI)       4621 non-null   float64
 3   (Low, ^NSEI)        4621 non-null   float64
 4   (Open, ^NSEI)       4621 non-null   float64
 5   (Volume, ^NSEI)     4621 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 252.7 KB


In [ ]:
# Official Yahoo Finance ticker
NIFTY_TICKER = "^NSEI"

In [ ]:
nifty = yf.download(
    NIFTY_TICKER,
    period="max",
    auto_adjust=False
)

[*********************100%***********************]  1 of 1 completed


In [ ]:
nifty.head()
nifty.tail()
nifty.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4621 entries, 2007-09-17 to 2026-07-21
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (Adj Close, ^NSEI)  4621 non-null   float64
 1   (Close, ^NSEI)      4621 non-null   float64
 2   (High, ^NSEI)       4621 non-null   float64
 3   (Low, ^NSEI)        4621 non-null   float64
 4   (Open, ^NSEI)       4621 non-null   float64
 5   (Volume, ^NSEI)     4621 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 252.7 KB


In [ ]:
print(nifty.columns)

MultiIndex([('Adj Close', '^NSEI'),
            (    'Close', '^NSEI'),
            (     'High', '^NSEI'),
            (      'Low', '^NSEI'),
            (     'Open', '^NSEI'),
            (   'Volume', '^NSEI')],
           names=['Price', 'Ticker'])


In [ ]:
def clean_yfinance_columns(df):
    """
    Convert yfinance MultiIndex columns into standard single-level columns.
    """

    if isinstance(df.columns, pd.MultiIndex):
        df.columns = df.columns.get_level_values(0)

    return df

In [ ]:
nifty = clean_yfinance_columns(nifty)

In [ ]:
print(nifty.columns)

Index(['Adj Close', 'Close', 'High', 'Low', 'Open', 'Volume'], dtype='object', name='Price')


In [ ]:
nifty.head()

Price,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
nifty.columns.name = None

In [ ]:
nifty.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
nifty.describe()

,Adj Close,Close,High,Low,Open,Volume
count,4621.000000,4621.000000,4621.000000,4621.000000,4621.000000,4.621000e+03
mean,11304.062558,11304.062558,11369.718739,11235.792422,11310.275667,2.172881e+05
std,6639.179926,6639.179926,6662.006898,6615.509421,6641.292294,2.065222e+05
min,2524.199951,2524.199951,2585.300049,2252.750000,2553.600098,0.000000e+00
25%,5718.799805,5718.799805,5755.000000,5677.899902,5717.100098,0.000000e+00
50%,8897.549805,8897.549805,8957.549805,8849.349609,8902.750000,1.940000e+05
75%,16584.300781,16584.300781,16649.199219,16447.900391,16545.250000,3.049000e+05
max,26328.550781,26328.550781,26373.199219,26210.050781,26333.699219,1.811000e+06


In [ ]:
import os

os.makedirs("data/raw", exist_ok=True)

nifty.to_csv("data/raw/nifty50_raw.csv", index=True)

In [ ]:
import os

print(os.listdir("data/raw"))

['nifty50_raw.csv']


In [ ]:
# Read the saved dataset

nifty_saved = pd.read_csv("data/raw/nifty50_raw.csv")

In [ ]:
nifty_saved.head()

,Date,Adj Close,Close,High,Low,Open,Volume
0,2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
1,2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2,2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
3,2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
4,2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
nifty_saved.shape

(4621, 7)

# Progress Summary

## Completed

- Downloaded NIFTY50 historical data
- Verified historical coverage (2007–2026)
- Standardized yfinance column names
- Validated data quality
- Saved raw dataset
- Verified saved CSV

## Dataset Summary

- Asset: NIFTY50
- Ticker: ^NSEI
- Source: Yahoo Finance (yfinance)
- Frequency: Daily
- Rows: 4621
- Columns: Date, Open, High, Low, Close, Adj Close, Volume

## Status

✅ Data collection pipeline successfully validated.

In [ ]:
def download_market_data(ticker, asset_name):
    """
    Downloads historical market data from Yahoo Finance,
    standardizes the columns, saves the raw CSV,
    and returns a cleaned DataFrame.

    Parameters
    ----------
    ticker : str
        Yahoo Finance ticker symbol.

    asset_name : str
        Name used when saving the dataset.

    Returns
    -------
    pandas.DataFrame
        Cleaned historical market data.
    """

    df = yf.download(
    ticker,
    start="2007-01-01",
    progress=False,
    auto_adjust=False
)


    df = clean_yfinance_columns(df)

    df.columns.name = None

    if df.empty:
        raise ValueError(f"No data downloaded for ticker: {ticker}")

    file_path = f"data/raw/{asset_name}_raw.csv"

    df.to_csv(file_path, index=True)

    print(f"Saved dataset to: {file_path}")

    return df

In [ ]:
print(download_market_data)

<function download_market_data at 0x7f8491840a40>


In [ ]:
test_nifty = download_market_data("^NSEI", "test_nifty")

Saved dataset to: data/raw/test_nifty_raw.csv


In [ ]:
test_nifty.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
test_nifty.shape

(4621, 6)

In [ ]:
vix = download_market_data("^INDIAVIX", "vix")
vix.shape

Saved dataset to: data/raw/vix_raw.csv


(4502, 6)

In [ ]:
sp500 = download_market_data("^GSPC", "sp500")
nasdaq = download_market_data("^IXIC", "nasdaq")
dow = download_market_data("^DJI", "dow")
usd_inr = download_market_data("INR=X", "usd_inr")
crude = download_market_data("CL=F", "crude")

Saved dataset to: data/raw/sp500_raw.csv
Saved dataset to: data/raw/nasdaq_raw.csv
Saved dataset to: data/raw/dow_raw.csv
Saved dataset to: data/raw/usd_inr_raw.csv
Saved dataset to: data/raw/crude_raw.csv


In [ ]:
sp500.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-03,1416.599976,1416.599976,1429.420044,1407.859985,1418.030029,3429160000
2007-01-04,1418.339966,1418.339966,1421.839966,1408.430054,1416.599976,3004460000
2007-01-05,1409.709961,1409.709961,1418.339966,1405.750000,1418.339966,2919400000
2007-01-08,1412.839966,1412.839966,1414.979980,1403.969971,1409.260010,2763340000
2007-01-09,1412.109985,1412.109985,1415.609985,1405.420044,1412.839966,3038380000


In [ ]:
vix.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2008-03-03,35.759998,35.759998,41.540001,28.490000,28.490000,0
2008-03-04,38.080002,38.080002,51.070000,30.590000,38.080002,0
2008-03-05,35.509998,35.509998,38.759998,27.950001,38.080002,0
2008-03-07,35.200001,35.200001,39.340000,27.049999,36.950001,0
2008-03-10,32.580002,32.580002,47.950001,25.330000,25.330000,0


In [ ]:
vix = download_market_data("^INDIAVIX", "vix")

vix.head()

Saved dataset to: data/raw/vix_raw.csv


,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2008-03-03,35.759998,35.759998,41.540001,28.490000,28.490000,0
2008-03-04,38.080002,38.080002,51.070000,30.590000,38.080002,0
2008-03-05,35.509998,35.509998,38.759998,27.950001,38.080002,0
2008-03-07,35.200001,35.200001,39.340000,27.049999,36.950001,0
2008-03-10,32.580002,32.580002,47.950001,25.330000,25.330000,0


In [ ]:
sp500 = download_market_data("^GSPC", "sp500")
nasdaq = download_market_data("^IXIC", "nasdaq")
dow = download_market_data("^DJI", "dow")
usd_inr = download_market_data("INR=X", "usd_inr")
crude = download_market_data("CL=F", "crude")

Saved dataset to: data/raw/sp500_raw.csv
Saved dataset to: data/raw/nasdaq_raw.csv
Saved dataset to: data/raw/dow_raw.csv
Saved dataset to: data/raw/usd_inr_raw.csv
Saved dataset to: data/raw/crude_raw.csv


In [ ]:
vix = download_market_data("^INDIAVIX", "vix")

Saved dataset to: data/raw/vix_raw.csv


In [ ]:
vix = vix.drop(columns=["Volume"], errors="ignore")

In [ ]:
sp500 = download_market_data("^GSPC", "sp500")
nasdaq = download_market_data("^IXIC", "nasdaq")
dow = download_market_data("^DJI", "dow")
usd_inr = download_market_data("INR=X", "usd_inr")
crude = download_market_data("CL=F", "crude")

Saved dataset to: data/raw/sp500_raw.csv
Saved dataset to: data/raw/nasdaq_raw.csv
Saved dataset to: data/raw/dow_raw.csv
Saved dataset to: data/raw/usd_inr_raw.csv
Saved dataset to: data/raw/crude_raw.csv


In [ ]:
sp500.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-03,1416.599976,1416.599976,1429.420044,1407.859985,1418.030029,3429160000
2007-01-04,1418.339966,1418.339966,1421.839966,1408.430054,1416.599976,3004460000
2007-01-05,1409.709961,1409.709961,1418.339966,1405.750000,1418.339966,2919400000
2007-01-08,1412.839966,1412.839966,1414.979980,1403.969971,1409.260010,2763340000
2007-01-09,1412.109985,1412.109985,1415.609985,1405.420044,1412.839966,3038380000


In [ ]:
nasdaq = download_market_data("^IXIC", "nasdaq")
dow = download_market_data("^DJI", "dow")
usd_inr = download_market_data("INR=X", "usd_inr")
crude = download_market_data("CL=F", "crude")

Saved dataset to: data/raw/nasdaq_raw.csv
Saved dataset to: data/raw/dow_raw.csv
Saved dataset to: data/raw/usd_inr_raw.csv
Saved dataset to: data/raw/crude_raw.csv


In [ ]:
nasdaq.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-03,2423.159912,2423.159912,2454.620117,2394.659912,2429.719971,2435280000
2007-01-04,2453.429932,2453.429932,2460.510010,2413.750000,2423.820068,2104210000
2007-01-05,2434.250000,2434.250000,2445.070068,2420.590088,2445.070068,2060360000
2007-01-08,2438.199951,2438.199951,2445.629883,2421.129883,2435.250000,1905620000
2007-01-09,2443.830078,2443.830078,2449.870117,2423.560059,2443.260010,2144160000


In [ ]:
usd_inr.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-01,44.215000,44.215000,44.223999,44.040001,44.220001,0
2007-01-02,44.122002,44.122002,44.220001,44.075001,44.212002,0
2007-01-03,44.111000,44.111000,44.410000,44.085999,44.119999,0
2007-01-04,44.095001,44.095001,44.480000,44.095001,44.119999,0
2007-01-05,44.139999,44.139999,44.330002,44.095001,44.095001,0


In [ ]:
dow = download_market_data("^DJI", "dow")
crude = download_market_data("CL=F", "crude")

Saved dataset to: data/raw/dow_raw.csv
Saved dataset to: data/raw/crude_raw.csv


In [ ]:
dow.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-03,12474.519531,12474.519531,12580.349609,12404.820312,12459.540039,327200000
2007-01-04,12480.690430,12480.690430,12510.410156,12403.860352,12473.160156,259060000
2007-01-05,12398.009766,12398.009766,12480.129883,12365.410156,12480.049805,235220000
2007-01-08,12423.490234,12423.490234,12445.919922,12337.370117,12392.009766,223500000
2007-01-09,12416.599609,12416.599609,12466.429688,12369.169922,12424.769531,225190000


In [ ]:
crude.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-01-02,61.049999,61.049999,61.549999,60.410000,60.980000,41686
2007-01-03,58.320000,58.320000,60.970001,57.720001,60.950001,248547
2007-01-04,55.590000,55.590000,58.580002,55.389999,58.349998,324010
2007-01-05,56.310001,56.310001,56.400002,54.900002,55.410000,250130
2007-01-08,56.090000,56.090000,57.720001,55.099998,56.240002,298727


In [ ]:
usd_inr = usd_inr.drop(columns=["Volume"], errors="ignore")

In [ ]:
print(test_nifty.shape)
print(vix.shape)
print(sp500.shape)
print(nasdaq.shape)
print(dow.shape)
print(usd_inr.shape)
print(crude.shape)

(4621, 6)
(4502, 5)
(4917, 6)
(4917, 6)
(4917, 6)
(5068, 5)
(4918, 6)


In [ ]:
vix = vix.drop(columns=["Volume"], errors="ignore")

In [ ]:
vix.head()

,Adj Close,Close,High,Low,Open
Date,,,,,
2008-03-03,35.759998,35.759998,41.540001,28.490000,28.490000
2008-03-04,38.080002,38.080002,51.070000,30.590000,38.080002
2008-03-05,35.509998,35.509998,38.759998,27.950001,38.080002
2008-03-07,35.200001,35.200001,39.340000,27.049999,36.950001
2008-03-10,32.580002,32.580002,47.950001,25.330000,25.330000


In [ ]:
def rename_market_columns(df, name):
    df = df.copy()

    df.columns = [
        f"{name}_{col.replace(' ', '_')}"
        for col in df.columns
    ]

    return df

In [ ]:
nifty = rename_market_columns(test_nifty, "NIFTY")
vix = rename_market_columns(vix, "VIX")
sp500 = rename_market_columns(sp500, "SP500")
nasdaq = rename_market_columns(nasdaq, "NASDAQ")
dow = rename_market_columns(dow, "DOW")
usd_inr = rename_market_columns(usd_inr, "USDINR")
crude = rename_market_columns(crude, "CRUDE")

In [ ]:
nifty.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
dir()


['DATA_FOLDER',
 'In',
 'NIFTY_TICKER',
 'Out',
 '_',
 '_16',
 '_18',
 '_19',
 '_23',
 '_24',
 '_28',
 '_29',
 '_30',
 '_32',
 '_33',
 '_34',
 '_39',
 '_41',
 '_42',
 '_44',
 '_45',
 '_49',
 '_52',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_i',
 '_i1',
 '_i10',
 '_i11',
 '_i12',
 '_i13',
 '_i14',
 '_i15',
 '_i16',
 '_i17',
 '_i18',
 '_i19',
 '_i2',
 '_i20',
 '_i21',
 '_i22',
 '_i23',
 '_i24',
 '_i25',
 '_i26',
 '_i27',
 '_i28',
 '_i29',
 '_i3',
 '_i30',
 '_i31',
 '_i32',
 '_i33',
 '_i34',
 '_i35',
 '_i36',
 '_i37',
 '_i38',
 '_i39',
 '_i4',
 '_i40',
 '_i41',
 '_i42',
 '_i43',
 '_i44',
 '_i45',
 '_i46',
 '_i47',
 '_i48',
 '_i49',
 '_i5',
 '_i50',
 '_i51',
 '_i52',
 '_i53',
 '_i6',
 '_i7',
 '_i8',
 '_i9',
 '_ih',
 '_ii',
 '_iii',
 '_oh',
 'clean_yfinance_columns',
 'crude',
 'dow',
 'download_market_data',
 'exit',
 'get_ipython',
 'nasdaq',
 'nifty',
 'nifty_saved',
 'np',
 'os',
 'pd',
 'quit',
 'renam

In [ ]:
test_nifty.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
nifty.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
test_nifty.shape, nifty.shape

((4621, 6), (4621, 6))

In [ ]:
test_nifty.equals(nifty)

False

In [ ]:
test_nifty.head()

,Adj Close,Close,High,Low,Open,Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
nifty.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0


In [ ]:
test_nifty.columns

Index(['Adj Close', 'Close', 'High', 'Low', 'Open', 'Volume'], dtype='object')

In [ ]:
nifty.columns

Index(['NIFTY_Adj_Close', 'NIFTY_Close', 'NIFTY_High', 'NIFTY_Low',
       'NIFTY_Open', 'NIFTY_Volume'],
      dtype='object')

In [ ]:
nifty

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume
Date,,,,,,
2007-09-17,4494.649902,4494.649902,4549.049805,4482.850098,4518.450195,0
2007-09-18,4546.200195,4546.200195,4551.799805,4481.549805,4494.100098,0
2007-09-19,4732.350098,4732.350098,4739.000000,4550.250000,4550.250000,0
2007-09-20,4747.549805,4747.549805,4760.850098,4721.149902,4734.850098,0
2007-09-21,4837.549805,4837.549805,4855.700195,4733.700195,4752.950195,0
...,...,...,...,...,...,...
2026-07-15,24078.500000,24078.500000,24220.349609,24010.550781,24085.849609,306900
2026-07-16,24072.750000,24072.750000,24186.500000,24050.000000,24142.099609,264300
2026-07-17,24334.300781,24334.300781,24367.300781,24099.050781,24127.599609,387800


In [ ]:
vix = rename_market_columns(vix, "VIX")
sp500 = rename_market_columns(sp500, "SP500")
nasdaq = rename_market_columns(nasdaq, "NASDAQ")
dow = rename_market_columns(dow, "DOW")
usd_inr = rename_market_columns(usd_inr, "USDINR")
crude = rename_market_columns(crude, "CRUDE")

In [ ]:
sp500.head()

,SP500_SP500_Adj_Close,SP500_SP500_Close,SP500_SP500_High,SP500_SP500_Low,SP500_SP500_Open,SP500_SP500_Volume
Date,,,,,,
2007-01-03,1416.599976,1416.599976,1429.420044,1407.859985,1418.030029,3429160000
2007-01-04,1418.339966,1418.339966,1421.839966,1408.430054,1416.599976,3004460000
2007-01-05,1409.709961,1409.709961,1418.339966,1405.750000,1418.339966,2919400000
2007-01-08,1412.839966,1412.839966,1414.979980,1403.969971,1409.260010,2763340000
2007-01-09,1412.109985,1412.109985,1415.609985,1405.420044,1412.839966,3038380000


In [ ]:
market_df = nifty.copy()

market_df = market_df.join(vix, how="inner")
market_df = market_df.join(sp500, how="inner")
market_df = market_df.join(nasdaq, how="inner")
market_df = market_df.join(dow, how="inner")
market_df = market_df.join(usd_inr, how="inner")
market_df = market_df.join(crude, how="inner")

In [ ]:
market_df.shape

(4348, 40)

In [ ]:
market_df.isnull().sum()

,0
NIFTY_Adj_Close,0
NIFTY_Close,0
NIFTY_High,0
NIFTY_Low,0
NIFTY_Open,0
NIFTY_Volume,0
VIX_VIX_Adj_Close,0
VIX_VIX_Close,0
VIX_VIX_High,0
VIX_VIX_Low,0


In [ ]:
market_df.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume,VIX_VIX_Adj_Close,VIX_VIX_Close,VIX_VIX_High,VIX_VIX_Low,...,USDINR_USDINR_Close,USDINR_USDINR_High,USDINR_USDINR_Low,USDINR_USDINR_Open,CRUDE_CRUDE_Adj_Close,CRUDE_CRUDE_Close,CRUDE_CRUDE_High,CRUDE_CRUDE_Low,CRUDE_CRUDE_Open,CRUDE_CRUDE_Volume
Date,,,,,,,,,,,,,,,,,,,,,
2008-03-03,4953.000000,4953.000000,5222.799805,4936.049805,5222.799805,0,35.759998,35.759998,41.540001,28.490000,...,40.254002,40.330002,40.160000,40.169998,102.449997,102.449997,103.949997,100.769997,101.629997,281850
2008-03-04,4864.250000,4864.250000,4976.700195,4812.950195,4958.549805,0,38.080002,38.080002,51.070000,30.590000,...,40.209000,40.360001,40.145000,40.279999,99.519997,99.519997,103.330002,98.870003,102.349998,366724
2008-03-05,4921.399902,4921.399902,4936.750000,4847.250000,4866.850098,0,35.509998,35.509998,38.759998,27.950001,...,40.174000,40.349998,40.125000,40.205002,104.519997,104.519997,104.949997,99.550003,99.879997,414308
2008-03-07,4771.600098,4771.600098,4918.299805,4672.250000,4918.299805,0,35.200001,35.200001,39.340000,27.049999,...,40.238998,40.630001,40.064999,40.064999,105.150002,105.150002,106.540001,103.910004,105.580002,321205
2008-03-10,4800.399902,4800.399902,4814.950195,4620.500000,4767.799805,0,32.580002,32.580002,47.950001,25.330000,...,40.313999,40.700001,40.305000,40.560001,107.900002,107.900002,108.209999,104.080002,105.250000,340551


In [ ]:
market_df.index.min(), market_df.index.max()

(Timestamp('2008-03-03 00:00:00'), Timestamp('2026-07-21 00:00:00'))

In [ ]:
market_df.columns = [
    col.replace("VIX_VIX", "VIX")
       .replace("SP500_SP500", "SP500")
       .replace("NASDAQ_NASDAQ", "NASDAQ")
       .replace("DOW_DOW", "DOW")
       .replace("USDINR_USDINR", "USDINR")
       .replace("CRUDE_CRUDE", "CRUDE")
    for col in market_df.columns
]

In [ ]:
market_df.columns

Index(['NIFTY_Adj_Close', 'NIFTY_Close', 'NIFTY_High', 'NIFTY_Low',
       'NIFTY_Open', 'NIFTY_Volume', 'VIX_Adj_Close', 'VIX_Close', 'VIX_High',
       'VIX_Low', 'VIX_Open', 'SP500_Adj_Close', 'SP500_Close', 'SP500_High',
       'SP500_Low', 'SP500_Open', 'SP500_Volume', 'NASDAQ_Adj_Close',
       'NASDAQ_Close', 'NASDAQ_High', 'NASDAQ_Low', 'NASDAQ_Open',
       'NASDAQ_Volume', 'DOW_Adj_Close', 'DOW_Close', 'DOW_High', 'DOW_Low',
       'DOW_Open', 'DOW_Volume', 'USDINR_Adj_Close', 'USDINR_Close',
       'USDINR_High', 'USDINR_Low', 'USDINR_Open', 'CRUDE_Adj_Close',
       'CRUDE_Close', 'CRUDE_High', 'CRUDE_Low', 'CRUDE_Open', 'CRUDE_Volume'],
      dtype='object')

In [ ]:
market_df["NIFTY_Return"] = market_df["NIFTY_Close"].pct_change()

market_df["SP500_Return"] = market_df["SP500_Close"].pct_change()

market_df["NASDAQ_Return"] = market_df["NASDAQ_Close"].pct_change()

market_df["DOW_Return"] = market_df["DOW_Close"].pct_change()

market_df["USDINR_Return"] = market_df["USDINR_Close"].pct_change()

market_df["CRUDE_Return"] = market_df["CRUDE_Close"].pct_change()

market_df["VIX_Return"] = market_df["VIX_Close"].pct_change()

In [ ]:
market_df.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume,VIX_Adj_Close,VIX_Close,VIX_High,VIX_Low,...,CRUDE_Low,CRUDE_Open,CRUDE_Volume,NIFTY_Return,SP500_Return,NASDAQ_Return,DOW_Return,USDINR_Return,CRUDE_Return,VIX_Return
Date,,,,,,,,,,,,,,,,,,,,,
2008-03-03,4953.000000,4953.000000,5222.799805,4936.049805,5222.799805,0,35.759998,35.759998,41.540001,28.490000,...,100.769997,101.629997,281850,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2008-03-04,4864.250000,4864.250000,4976.700195,4812.950195,4958.549805,0,38.080002,38.080002,51.070000,30.590000,...,98.870003,102.349998,366724,-0.017918,-0.003448,0.000744,-0.003679,-0.001118,-0.028599,0.064877
2008-03-05,4921.399902,4921.399902,4936.750000,4847.250000,4866.850098,0,35.509998,35.509998,38.759998,27.950001,...,99.550003,99.879997,414308,0.011749,0.005238,0.005544,0.003372,-0.000870,0.050241,-0.067490
2008-03-07,4771.600098,4771.600098,4918.299805,4672.250000,4918.299805,0,35.200001,35.200001,39.340000,27.049999,...,103.910004,105.580002,321205,-0.030438,-0.030239,-0.026540,-0.029482,0.001618,0.006028,-0.008730
2008-03-10,4800.399902,4800.399902,4814.950195,4620.500000,4767.799805,0,32.580002,32.580002,47.950001,25.330000,...,104.080002,105.250000,340551,0.006036,-0.015463,-0.019503,-0.012909,0.001864,0.026153,-0.074432


In [ ]:
market_df = market_df.dropna()

In [ ]:
market_df.shape

(4347, 47)

In [ ]:
market_df.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume,VIX_Adj_Close,VIX_Close,VIX_High,VIX_Low,...,CRUDE_Low,CRUDE_Open,CRUDE_Volume,NIFTY_Return,SP500_Return,NASDAQ_Return,DOW_Return,USDINR_Return,CRUDE_Return,VIX_Return
Date,,,,,,,,,,,,,,,,,,,,,
2008-03-04,4864.250000,4864.250000,4976.700195,4812.950195,4958.549805,0,38.080002,38.080002,51.070000,30.590000,...,98.870003,102.349998,366724,-0.017918,-0.003448,0.000744,-0.003679,-0.001118,-0.028599,0.064877
2008-03-05,4921.399902,4921.399902,4936.750000,4847.250000,4866.850098,0,35.509998,35.509998,38.759998,27.950001,...,99.550003,99.879997,414308,0.011749,0.005238,0.005544,0.003372,-0.000870,0.050241,-0.067490
2008-03-07,4771.600098,4771.600098,4918.299805,4672.250000,4918.299805,0,35.200001,35.200001,39.340000,27.049999,...,103.910004,105.580002,321205,-0.030438,-0.030239,-0.026540,-0.029482,0.001618,0.006028,-0.008730
2008-03-10,4800.399902,4800.399902,4814.950195,4620.500000,4767.799805,0,32.580002,32.580002,47.950001,25.330000,...,104.080002,105.250000,340551,0.006036,-0.015463,-0.019503,-0.012909,0.001864,0.026153,-0.074432
2008-03-11,4865.899902,4865.899902,4888.500000,4732.850098,4796.299805,0,32.689999,32.689999,37.540001,29.750000,...,106.610001,107.910004,349815,0.013645,0.037130,0.039837,0.035490,-0.003448,0.007878,0.003376


In [ ]:
market_df = market_df.dropna()
market_df.shape

(4347, 47)

In [ ]:
market_df.head()

,NIFTY_Adj_Close,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume,VIX_Adj_Close,VIX_Close,VIX_High,VIX_Low,...,CRUDE_Low,CRUDE_Open,CRUDE_Volume,NIFTY_Return,SP500_Return,NASDAQ_Return,DOW_Return,USDINR_Return,CRUDE_Return,VIX_Return
Date,,,,,,,,,,,,,,,,,,,,,
2008-03-04,4864.250000,4864.250000,4976.700195,4812.950195,4958.549805,0,38.080002,38.080002,51.070000,30.590000,...,98.870003,102.349998,366724,-0.017918,-0.003448,0.000744,-0.003679,-0.001118,-0.028599,0.064877
2008-03-05,4921.399902,4921.399902,4936.750000,4847.250000,4866.850098,0,35.509998,35.509998,38.759998,27.950001,...,99.550003,99.879997,414308,0.011749,0.005238,0.005544,0.003372,-0.000870,0.050241,-0.067490
2008-03-07,4771.600098,4771.600098,4918.299805,4672.250000,4918.299805,0,35.200001,35.200001,39.340000,27.049999,...,103.910004,105.580002,321205,-0.030438,-0.030239,-0.026540,-0.029482,0.001618,0.006028,-0.008730
2008-03-10,4800.399902,4800.399902,4814.950195,4620.500000,4767.799805,0,32.580002,32.580002,47.950001,25.330000,...,104.080002,105.250000,340551,0.006036,-0.015463,-0.019503,-0.012909,0.001864,0.026153,-0.074432
2008-03-11,4865.899902,4865.899902,4888.500000,4732.850098,4796.299805,0,32.689999,32.689999,37.540001,29.750000,...,106.610001,107.910004,349815,0.013645,0.037130,0.039837,0.035490,-0.003448,0.007878,0.003376


In [ ]:
adj_columns = [col for col in market_df.columns if "Adj_Close" in col]

market_df = market_df.drop(columns=adj_columns)

market_df.shape

(4347, 40)

In [ ]:
market_df["Target"] = (
    market_df["NIFTY_Close"].shift(-1) > market_df["NIFTY_Close"]
).astype(int)

In [ ]:
market_df[["NIFTY_Close", "Target"]].head()

,NIFTY_Close,Target
Date,,
2008-03-04,4864.250000,1
2008-03-05,4921.399902,0
2008-03-07,4771.600098,1
2008-03-10,4800.399902,1
2008-03-11,4865.899902,1


In [ ]:
market_df = market_df.dropna()

In [ ]:
market_df.shape

(4347, 41)

In [ ]:
market_df["Target"].value_counts()

,count
Target,
1,2307
0,2040


In [ ]:
market_df[["NIFTY_Close", "Target"]].head()

,NIFTY_Close,Target
Date,,
2008-03-04,4864.250000,1
2008-03-05,4921.399902,0
2008-03-07,4771.600098,1
2008-03-10,4800.399902,1
2008-03-11,4865.899902,1


In [ ]:
X = market_df.drop("Target", axis=1)
y = market_df["Target"]

In [ ]:
market_df.head()

,NIFTY_Close,NIFTY_High,NIFTY_Low,NIFTY_Open,NIFTY_Volume,VIX_Close,VIX_High,VIX_Low,VIX_Open,SP500_Close,...,CRUDE_Open,CRUDE_Volume,NIFTY_Return,SP500_Return,NASDAQ_Return,DOW_Return,USDINR_Return,CRUDE_Return,VIX_Return,Target
Date,,,,,,,,,,,,,,,,,,,,,
2008-03-04,4864.250000,4976.700195,4812.950195,4958.549805,0,38.080002,51.070000,30.590000,38.080002,1326.750000,...,102.349998,366724,-0.017918,-0.003448,0.000744,-0.003679,-0.001118,-0.028599,0.064877,1
2008-03-05,4921.399902,4936.750000,4847.250000,4866.850098,0,35.509998,38.759998,27.950001,38.080002,1333.699951,...,99.879997,414308,0.011749,0.005238,0.005544,0.003372,-0.000870,0.050241,-0.067490,0
2008-03-07,4771.600098,4918.299805,4672.250000,4918.299805,0,35.200001,39.340000,27.049999,36.950001,1293.369995,...,105.580002,321205,-0.030438,-0.030239,-0.026540,-0.029482,0.001618,0.006028,-0.008730,1
2008-03-10,4800.399902,4814.950195,4620.500000,4767.799805,0,32.580002,47.950001,25.330000,25.330000,1273.369995,...,105.250000,340551,0.006036,-0.015463,-0.019503,-0.012909,0.001864,0.026153,-0.074432,1
2008-03-11,4865.899902,4888.500000,4732.850098,4796.299805,0,32.689999,37.540001,29.750000,30.719999,1320.650024,...,107.910004,349815,0.013645,0.037130,0.039837,0.035490,-0.003448,0.007878,0.003376,1


In [ ]:
market_df.to_csv("market_data_processed.csv")

In [ ]:
from google.colab import files

files.download("market_data_processed.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
print(market_df.index.min())
print(market_df.index.max())

2008-03-04 00:00:00
2026-07-21 00:00:00
